vault/login_mfa.go

```diff
@@ -647,6 +674,87 @@ func (b *LoginMFABackend) loginMFAMethodExistenceCheck(eConfig *mfa.MFAEnforceme
 	return aggErr.ErrorOrNil()
 }
 
+// sanitizeMethodNameSameNamespace returns the name without the namespace.
+// The method name could contain the namespace path.
+// If the namespace is specified in the name, check if the namespace
+// matches the request namespace. If so, trim the namespace path and
+// return the name. If it does not match the request namespace,
+// return an error. If the namespace path is not included in the name,
+// nothing happens.
+func (b *LoginMFABackend) sanitizeMethodNameSameNamespace(ns *namespace.Namespace, name string) (string, error) {
+	if name == "" {
+		return "", nil
+	}
+
+	if ns == nil {
+		return "", fmt.Errorf("namespace cannot be nil")
+	}
+
+	if !strings.Contains(name, "/") {
+		return name, nil
+	}
+
+	// method name
+	methodName := path.Base(name)
+	// namespace path in method name
+	methodNamespacePath := strings.TrimSuffix(name, methodName)
+
+	if ns.Path != methodNamespacePath {
+		return "", fmt.Errorf("invalid underlying namespace %s for method name %s", methodNamespacePath, methodName)
+	}
+
+	return methodName, nil
+}
+
+// sanitizeMFACredsWithLoginEnforcementMethodIDs updates the MFACred map
+// looping through the matched login enforcement configurations, and
+// replacing MFA method names with MFA method IDs
+func (b *LoginMFABackend) sanitizeMFACredsWithLoginEnforcementMethodIDs(ctx context.Context, mfaCredsMap logical.MFACreds, mfaMethodIDs []string) (logical.MFACreds, error) {
+	sanitizedMfaCreds := make(logical.MFACreds, 0)
+	var multiError *multierror.Error
+	for _, methodID := range mfaMethodIDs {
+		val, ok := mfaCredsMap[methodID]
+		if ok {
+			sanitizedMfaCreds[methodID] = val
+			continue
+		}
+		mConfig, err := b.MemDBMFAConfigByID(methodID)
+		if err != nil {
+			return nil, err
+		}
+		// method name in the MFACredsMap could be just the method full name,
+		// i.e., namespace path+name, or just the name without namespace.
+		// let's check for both cases.
+		val, ok = mfaCredsMap[mConfig.Name]
+		if ok {
+			sanitizedMfaCreds[mConfig.ID] = val
+		} else {
+			configNS, err := NamespaceByID(ctx, mConfig.NamespaceID, b.Core)
+			if err != nil {
+				return nil, err
+			}
+			if configNS != nil {
+				val, ok = mfaCredsMap[configNS.Path+mConfig.Name]
+				if ok {
+					sanitizedMfaCreds[mConfig.ID] = val
+				} else {
+					multiError = multierror.Append(multiError, fmt.Errorf("failed to find MFA credentials associated with an MFA method ID %v, method name %v", methodID, mConfig.Name))
+				}
+			} else {
+				multiError = multierror.Append(multiError, fmt.Errorf("failed to find the namespace associated with an MFA method"))
+			}
+		}
+	}
+
+	// we don't need to find every MFA method identifiers in the MFA header
+	// So, don't return errors if that is the case.
```

Comment:
Is there a preference on which ones we should find or does it just matter that we found some at all?

Reply:
Got it, thanks